rasa/core/test.py

```diff
@@ -105,26 +104,104 @@ def has_prediction_target_mismatch(self) -> bool:
 
     def serialise(self) -> Tuple[List[Text], List[Text]]:
         """Turn targets and predictions to lists of equal size for sklearn."""
+        # sklearn does not cope with lists of unequal size, nor None values
 
-        targets = (
-            self.action_targets
-            + self.intent_targets
-            + [
-                TrainingDataWriter.generate_entity(gold.get("text"), gold)
-                for gold in self.entity_targets
-            ]
+        texts = sorted(
+            list(
+                set(
+                    [e.get("text") for e in self.entity_targets]
+                    + [e.get("text") for e in self.entity_predictions]
+                )
+            )
         )
+
+        entity_targets_fixed = []
+        entity_predictions_fixed = []
+
+        for text in texts:
+            # sort the entities of this sentence to compare them directly
+            entity_targets = sorted(
+                filter(lambda x: x.get("text") == text, self.entity_targets),
+                key=lambda x: x.get("start"),
+            )
+            entity_predictions = sorted(
+                filter(lambda x: x.get("text") == text, self.entity_predictions),
+                key=lambda x: x.get("start"),
+            )
+
+            i_pred = 0
+            i_target = 0
+
+            def compare_entities():
```

Comment:
I would move this method out of the method `serialise` and make it private.

Reply:
Also can you please add some comments to the methods to make it even more clear what cases we check.